### PR TITLE
Reports widget extraction workflow update

### PR DESCRIPTION
**Pull request by Simonas-A**
Extraction workflow for mappings updated. Old Extraction component removed. Tests refactored.

### CODE REVIEW

File: packages/itwin/reports-config-widget/src/widget/components/BulkExtractor.ts [hunk 1]
```diff
@@ -21,23 +21,30 @@ export default class BulkExtractor {
...
   private _timeFetched = new Date();
-  private _reportIds: string[];
-
-  constructor(apiConfig: ReportsApiConfig, reportIds: string[]) {
+  private _iModelRun = new Map<string, string>();
+  private _iModelToast = new Map<string, boolean>();
+  private _successfulExtractionToast: (iModelName: string, odataFeedUrl: string) => void;
+  private _failedExtractionToast: (iModelName: string) => void;
+
+  constructor(apiConfig: ReportsApiConfig,
```
Comment (MindaugasButkus) [inline packages/itwin/reports-config-widget/src/widget/components/BulkExtractor.ts:31]: Move `apiConfig` to new line and the ctor args closing parenthesis to a new line. Now it is hard to see where the ctor args end.

File: packages/itwin/reports-config-widget/src/widget/components/BulkExtractor.ts [hunk 2]
```diff
@@ -95,7 +143,7 @@ export default class BulkExtractor {
     return ExtractionStates.Failed;
   }
 
-  private async getSingleState(runId: string, accessToken: string): Promise<ExtractorState> {
+  private async getState(runId: string, accessToken: string): Promise<ExtractorState> {
```
Comment (MindaugasButkus) [inline packages/itwin/reports-config-widget/src/widget/components/BulkExtractor.ts:146]: You can take `accessToken` with `await this._accessToken()`.

File: packages/itwin/reports-config-widget/src/widget/components/ReportsContainer.scss [hunk 3]
```diff
@@ -5,9 +5,6 @@
 @import "~@itwin/itwinui-css/scss/variables";
 
 .reports-container {
-  position: absolute;
-  right: 0;
-  left: 0;
```
Comment (Guillar1) [inline packages/itwin/reports-config-widget/src/widget/components/ReportsContainer.scss:10]: You're going to have to bump up the minimum iTwin.js version to ^3.3.5. Removing this workaround will break the styling for any version below 3.3.5 as they don't include this fix: https://github.com/iTwin/itwinjs-core/pull/4266

File: packages/itwin/reports-config-widget/src/widget/components/ReportsContainer.scss [hunk 4]
```diff
@@ -5,9 +5,6 @@
 @import "~@itwin/itwinui-css/scss/variables";
 
 .reports-container {
```
Comment (MindaugasButkus) [inline packages/itwin/reports-config-widget/src/widget/components/ReportsContainer.scss:7]: Prefix all style classes with `rcw-`.

File: packages/itwin/reports-config-widget/src/widget/components/Reports.tsx [hunk 5]
```diff
@@ -176,7 +186,7 @@ export const Reports = () => {
                 onClick={updateDatasets}
                 disabled={selectedReportIds.length === 0}
               >
-                <SvgRefresh />
+                < SvgPlay />
```
Comment (MindaugasButkus) [inline packages/itwin/reports-config-widget/src/widget/components/Reports.tsx:189]: Formatting - remove spaces after the opening `<`.

File: packages/itwin/reports-config-widget/src/widget/components/BulkExtractor.ts [hunk 6]
```diff
@@ -49,34 +56,75 @@ export default class BulkExtractor {
...
+    this._reportStates = stateByReportId;
+    this._iModelStates = stateByIModelId;
   }
 
-  public getState(reportId: string): ExtractionStates {
+  public async getReportState(reportId: string): Promise<ExtractionStates> {
     if ((new Date().getTime() - this._timeFetched.getTime()) > STATUS_CHECK_INTERVAL) {
       this._timeFetched = new Date();
-      this.fetchStates().catch((e) =>
+      await this.fetchStates().catch((e) =>
```
Comment (MindaugasButkus) [inline packages/itwin/reports-config-widget/src/widget/components/BulkExtractor.ts:87]: Change to `try/catch` instead of `.catch(() => ...)`

File: packages/itwin/reports-config-widget/src/widget/components/BulkExtractor.ts [hunk 7]
```diff
@@ -49,34 +56,75 @@ export default class BulkExtractor {
...
+      this._successfulExtractionToast(iModelName, odataFeedUrl);
+      this._iModelToast.set(iModelId, true);
+    } else if (state === ExtractionStates.Failed && !this._iModelToast.get(iModelId)) {
+      this._failedExtractionToast(iModelName);
+      this._iModelToast.set(iModelId, true);
+    }
+    return state;
+  }
+
+  public clearReportJob(reportId: string): void {
```
Comment (MindaugasButkus) [inline packages/itwin/reports-config-widget/src/widget/components/BulkExtractor.ts:118]: Looks like this method clears a lot more than just the reportJob. Why is this needed?

File: packages/itwin/reports-config-widget/src/widget/components/BulkExtractor.ts [hunk 8]
```diff
@@ -21,23 +21,30 @@ export default class BulkExtractor {
   private _extractionClientApi: ExtractionClient;
   private _accessToken: () => Promise<string>;
   private _reportStates = new Map<string, ExtractionStates>();
+  private _iModelStates = new Map<string, ExtractionStates>();
   private _timeFetched = new Date();
-  private _reportIds: string[];
-
-  constructor(apiConfig: ReportsApiConfig, reportIds: string[]) {
+  private _iModelRun = new Map<string, string>();
+  private _iModelToast = new Map<string, boolean>();
```
Comment (MindaugasButkus) [inline packages/itwin/reports-config-widget/src/widget/components/BulkExtractor.ts:27]: I think `_iModelToast` could be changed to a `Set<string>`. You could also rename this to `_toastediModels` or something like that, because it keeps track of which iModels already appeared in a toast message.

File: packages/itwin/reports-config-widget/src/widget/components/ReportMappings.tsx [hunk 9]
```diff
@@ -342,6 +292,41 @@ export const ReportMappings = ({ report, goBack }: ReportMappingsProps) => {
         }}
         refresh={refresh}
       />
+      <div id="action">
```
Comment (MindaugasButkus) [inline packages/itwin/reports-config-widget/src/widget/components/ReportMappings.tsx:295]: Is this `id` needed? Is this `div` element needed?

File: packages/itwin/reports-config-widget/src/widget/components/ReportMappings.tsx [hunk 10]
```diff
@@ -342,6 +292,41 @@ export const ReportMappings = ({ report, goBack }: ReportMappingsProps) => {
...
+        <div className="rcw-action-panel">
+          {isLoading && <LoadingSpinner />}
+          <Button
+            disabled={isLoading}
+            styleType="high-visibility"
+            id="save-app"
```
Comment (MindaugasButkus) [inline packages/itwin/reports-config-widget/src/widget/components/ReportMappings.tsx:301]: Is this `id` needed?

File: packages/itwin/reports-config-widget/src/widget/components/ReportMappings.tsx [hunk 11]
```diff
@@ -342,6 +292,41 @@ export const ReportMappings = ({ report, goBack }: ReportMappingsProps) => {
...
+            }}
+          >
+            {ReportsConfigWidget.localization.getLocalizedString(
+              "ReportsConfigWidget:UpdateAllDatasets"
+            )}
+          </Button>
+          <Button
+            styleType="default"
+            type="button"
+            id="cancel"
```
Comment (MindaugasButkus) [inline packages/itwin/reports-config-widget/src/widget/components/ReportMappings.tsx:320]: Is this `id` needed?

File: packages/itwin/reports-config-widget/src/widget/components/ExtractionToast.tsx [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+interface ExtractionToastProps {
+  iModelName: string;
+  odataFeedUrl?: string;
+}
+
+export const SuccessfulExtractionToast = ({ iModelName, odataFeedUrl }: ExtractionToastProps) => {
+  return (
+    <div>
+      <Text>{ReportsConfigWidget.localization.getLocalizedString("ReportsConfigWidget:ExtractionSuccess")}{iModelName} </Text>
+      <a href="#" onClick={async () => { await navigator.clipboard.writeText(odataFeedUrl!); }}>{ReportsConfigWidget.localization.getLocalizedString("ReportsConfigWidget:CopyODataUrl")}</a>
```
Comment (MindaugasButkus) [inline packages/itwin/reports-config-widget/src/widget/components/ExtractionToast.tsx:18]: Does `#` appear in the url when you click on this link?

File: packages/itwin/reports-config-widget/src/widget/components/ExtractionToast.tsx [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+import React from "react";
+import { Text } from "@itwin/itwinui-react";
+import { ReportsConfigWidget } from "../../ReportsConfigWidget";
+
+interface ExtractionToastProps {
+  iModelName: string;
+  odataFeedUrl?: string;
```
Comment (MindaugasButkus) [inline packages/itwin/reports-config-widget/src/widget/components/ExtractionToast.tsx:11]: Split into 2 interfaces. One with `odataFeedUrl` and another without it. Looks like `odataFeedUrl` is not optional for `SuccessfulExtractionToast`.

File: packages/itwin/reports-config-widget/src/widget/components/ReportMappingHorizontalTile.tsx [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+          {extractionState === ExtractionStates.None ?
+            (
+              <IconButton
+                styleType="borderless"
+                title={ReportsConfigWidget.localization.getLocalizedString(
+                  "ReportsConfigWidget:UpdateDataset"
+                )}
+                onClick={async () => {
+                  setExtractionState(ExtractionStates.Starting);
+                  await props.bulkExtractor.runIModelExtraction(props.mapping.imodelId).catch((e) => {
```
Comment (MindaugasButkus) [inline packages/itwin/reports-config-widget/src/widget/components/ReportMappingHorizontalTile.tsx:89]: Use `try/catch`.

File: packages/itwin/reports-config-widget/src/widget/components/ExtractionToast.tsx (hunk 12 above)
Comment (Guillar1) [inline packages/itwin/reports-config-widget/src/widget/components/ExtractionToast.tsx:18]: I feel like this line is too long. It's confusing to read at first.

Comment (MindaugasButkus) [reply packages/itwin/reports-config-widget/src/widget/components/ExtractionToast.tsx:18]: Move the `onClick` handler to a const and the contents of the `<a>` tag on a separate line. That should help with readability.

File: packages/itwin/reports-config-widget/src/widget/components/BulkExtractor.ts (hunk 1 above)
Comment (Simonas-A) [reply packages/itwin/reports-config-widget/src/widget/components/BulkExtractor.ts]: `apiConfig` moved to a new line

File: packages/itwin/reports-config-widget/src/widget/components/BulkExtractor.ts (hunk 2 above)
Comment (Simonas-A) [reply packages/itwin/reports-config-widget/src/widget/components/BulkExtractor.ts]: `AccessToken` is now awaited inside the method

File: packages/itwin/reports-config-widget/src/widget/components/ReportsContainer.scss (hunk 4 above)
Comment (Simonas-A) [reply packages/itwin/reports-config-widget/src/widget/components/ReportsContainer.scss]: All styles have been prefixed

File: packages/itwin/reports-config-widget/src/widget/components/Reports.tsx (hunk 5 above)
Comment (Simonas-A) [reply packages/itwin/reports-config-widget/src/widget/components/Reports.tsx]: Space removed

File: packages/itwin/reports-config-widget/src/widget/components/BulkExtractor.ts (hunk 6 above)
Comment (Simonas-A) [reply packages/itwin/reports-config-widget/src/widget/components/BulkExtractor.ts]: Unnecessary `try/catch` removed

File: packages/itwin/reports-config-widget/src/widget/components/BulkExtractor.ts (hunk 7 above)
Comment (Simonas-A) [reply packages/itwin/reports-config-widget/src/widget/components/BulkExtractor.ts]: `BulkExtractor` refactored so this method is no longer needed and has been removed.

File: packages/itwin/reports-config-widget/src/widget/components/BulkExtractor.ts (hunk 8 above)
Comment (Simonas-A) [reply packages/itwin/reports-config-widget/src/widget/components/BulkExtractor.ts]: `_iModelToast` changed to `Set<string>`

File: packages/itwin/reports-config-widget/src/widget/components/BulkExtractor.ts [hunk 15]
```diff
@@ -139,13 +189,30 @@ export default class BulkExtractor {
...
       );
+      this._iModelToast.set(iModelId, false);
       return response.id;
     } catch (error: any) {
       handleError(error.status);
     }
     return undefined;
   }
 
+  public async runIModelExtraction(iModelId: string): Promise<void> {
```
[the remark this answers is not in the document]
Comment (Simonas-A) [reply packages/itwin/reports-config-widget/src/widget/components/BulkExtractor.ts:164]: Method has been simplified

File: packages/itwin/reports-config-widget/src/widget/components/ReportMappings.tsx (hunk 9 above)
Comment (Simonas-A) [reply packages/itwin/reports-config-widget/src/widget/components/ReportMappings.tsx]: `id` has been removed

File: packages/itwin/reports-config-widget/src/widget/components/ReportMappings.tsx (hunk 10 above)
Comment (Simonas-A) [reply packages/itwin/reports-config-widget/src/widget/components/ReportMappings.tsx]: `id` has been removed

File: packages/itwin/reports-config-widget/src/widget/components/ReportMappings.tsx (hunk 11 above)
Comment (Simonas-A) [reply packages/itwin/reports-config-widget/src/widget/components/ReportMappings.tsx]: `id` has been removed

File: packages/itwin/reports-config-widget/src/widget/components/ExtractionToast.tsx (hunk 12 above)
Comment (Simonas-A) [reply packages/itwin/reports-config-widget/src/widget/components/ExtractionToast.tsx]: `e.preventDefault();` has been added. The `#` no longer appears in the link

File: packages/itwin/reports-config-widget/src/widget/components/ExtractionToast.tsx (hunk 13 above)
Comment (Simonas-A) [reply packages/itwin/reports-config-widget/src/widget/components/ExtractionToast.tsx]: Interface has been split into two interfaces.

File: packages/itwin/reports-config-widget/src/widget/components/ReportMappingHorizontalTile.tsx (hunk 14 above)
Comment (Simonas-A) [reply packages/itwin/reports-config-widget/src/widget/components/ReportMappingHorizontalTile.tsx]: Unnecessary `try/catch` has been removed

File: packages/itwin/reports-config-widget/src/widget/components/ExtractionToast.tsx (hunk 12 above)
Comment (Simonas-A) [reply packages/itwin/reports-config-widget/src/widget/components/ExtractionToast.tsx]: `onClick` handler has been moved to a const. `<a>` tag contents separated

File: packages/itwin/reports-config-widget/src/widget/components/ReportMappingHorizontalTile.tsx [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+            onClick={() => {
+              props.onClickDelete();
+            }}
+            disabled={jobStarted}
+          >
+            <SvgDelete />
+          </IconButton>
+        </div >
+      )}
+    ></HorizontalTile >
```
Comment (MindaugasButkus) [inline packages/itwin/reports-config-widget/src/widget/components/ReportMappingHorizontalTile.tsx:126]:
```suggestion
    />
```

File: packages/itwin/reports-config-widget/src/test/ReportMappings.test.tsx [hunk 17]
```diff
@@ -419,289 +384,40 @@ describe("Report Mappings View", () => {
...
-    const modal = screen.getByRole("dialog");
-    expect(screen.getByRole("dialog")).toBeInTheDocument();
-
-    const withinModal = within(modal);
-    expect(withinModal.getByText(/addmappings/i)).toBeInTheDocument();
+    mockGetMappings.mockReturnValueOnce(mockMappings.map((m: MappingSingle) => m.mapping));
 
-    const addButton = withinModal.getByRole("button", {
-      name: /add/i,
+    await act(async () => {
```
Comment (MindaugasButkus) [inline packages/itwin/reports-config-widget/src/test/ReportMappings.test.tsx:408]: What does this `act` do?

File: packages/itwin/reports-config-widget/src/test/ReportMappings.test.tsx [hunk 18]
```diff
@@ -419,289 +384,40 @@ describe("Report Mappings View", () => {
     ).toBeInTheDocument();
   });
 
-  it("remove a report mapping", async () => {
```
Comment (MindaugasButkus) [inline packages/itwin/reports-config-widget/src/test/ReportMappings.test.tsx:422]: Why did you remove the `remove a report mapping` test?

File: packages/itwin/reports-config-widget/src/test/DeleteModal.test.tsx [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+const mockITwinId = faker.datatype.uuid();
+const mockIModelId1 = faker.datatype.uuid();
+
+const connectionMock = moq.Mock.ofType<IModelConnection>();
+const selectionManagerMock = moq.Mock.ofType<SelectionManager>();
+const selectionScopesManagerMock = moq.Mock.ofType<SelectionScopesManager>();
+
+beforeAll(async () => {
+  // This is required by the i18n module within iTwin.js
+  (global as any).XMLHttpRequest = require("xmlhttprequest").XMLHttpRequest; // eslint-disable-line @typescript-eslint/no-var-requires
```
Comment (MindaugasButkus) [inline packages/itwin/reports-config-widget/src/test/DeleteModal.test.tsx:41]: Why is this needed? All requests should be done through mocked clients.

File: packages/itwin/reports-config-widget/src/test/AddMappingModal.test.tsx [hunk 20]
```diff
@@ -0,0 +1,316 @@
...
+    getMappings: mockGetMappings,
+  })),
+  ReportsClient: jest.fn().mockImplementation(() => ({
+    createReportMapping: mockCreateReportMapping,
+  })),
+}));
+
+beforeAll(async () => {
+  // This is required by the i18n module within iTwin.js
+  (global as any).XMLHttpRequest = require("xmlhttprequest").XMLHttpRequest; // eslint-disable-line @typescript-eslint/no-var-requires
```
Comment (MindaugasButkus) [inline packages/itwin/reports-config-widget/src/test/AddMappingModal.test.tsx:242]: This should not be needed since all requests should go through mocked clients.

File: packages/itwin/reports-config-widget/src/test/ReportMappings.test.tsx (hunk 17 above)
Comment (Simonas-A) [reply packages/itwin/reports-config-widget/src/test/ReportMappings.test.tsx:408]: `act` is required to wait for all asynchronous changes in the component to be completed.

Comment (MindaugasButkus) [reply packages/itwin/reports-config-widget/src/test/ReportMappings.test.tsx:408]: From https://reactjs.org/docs/testing-recipes.html#act :
```
When writing UI tests, tasks like rendering, user events, or data fetching can be considered as “units” of interaction with a user interface. react-dom/test-utils provides a helper called [act()](https://reactjs.org/docs/test-utils.html#act) that makes sure all updates related to these “units” have been processed and applied to the DOM before you make any assertions.
```

I guess it would be best to wrap any "clicks" or renders with `act` for better consistency.

Comment (MindaugasButkus) [reply packages/itwin/reports-config-widget/src/test/ReportMappings.test.tsx:408]: I think anything that causes a re-render should be wrapped in `act`.

File: packages/itwin/reports-config-widget/src/test/ReportMappings.test.tsx (hunk 18 above)
Comment (Simonas-A) [reply packages/itwin/reports-config-widget/src/test/ReportMappings.test.tsx:422]: `remove a report mapping` test has been moved to `ReportMappingHorizontalTile` tests, since the only way to trigger deletion of Report Mapping is through `HorizontalTile`.

File: packages/itwin/reports-config-widget/src/test/ReportMappings.test.tsx (hunk 17 above)
Comment (Simonas-A) [reply packages/itwin/reports-config-widget/src/test/ReportMappings.test.tsx:408]: `click` function provided by `react-testing-library` is already wrapped in `act`. Should all initial component renders be wrapped in `act`?

Comment (MindaugasButkus) [reply packages/itwin/reports-config-widget/src/test/ReportMappings.test.tsx:408]: Yes, I think it would be good to wrap them.

File: packages/itwin/reports-config-widget/src/widget/components/ReportsContainer.scss (hunk 3 above)
Comment (MindaugasButkus) [reply packages/itwin/reports-config-widget/src/widget/components/ReportsContainer.scss:10]: Change was reverted to avoid the problems that package upgrades might bring.

**Review by MindaugasButkus — APPROVED** [review]

**Comment by MindaugasButkus** [comment]
@aruniverse please review this PR.

**Review by aruniverse — APPROVED** [review]
approving lockfile & pkg.json changes only
